<a href="https://colab.research.google.com/github/yunmokoo/SPY_MARS_LSTM/blob/main/SPY.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install requests > make_outputrerquest.txt 2>&1
!pip install yfinance > make_outputyfinance.txt 2>&1



In [ ]:
!wget http://prdownloads.sourceforge.net/ta-lib/ta-lib-0.4.0-src.tar.gz > wget_output.txt 2>&1
!tar -xzvf ta-lib-0.4.0-src.tar.gz > tar_output.txt 2>&1
%cd ta-lib
!./configure --prefix=/usr > config_output.txt 2>&1
!make > make_outputtalib.txt 2>&1
!make install > makeInstall_output.txt 2>&1
!pip install TA-Lib > pipInstall_output.txt 2>&1

/content/ta-lib


In [ ]:
import requests
import json
import yfinance as yf
import talib

import tensorflow as tf
import numpy as np
import pandas as pd
import torch
import matplotlib as plt

From TA-lib...

1. MA                  All Moving Average
2. BBANDS              Bollinger Bands
3. MACD                Moving Average Convergence/Divergence
4. SMA                 Simple Moving Average
5. EMA                 Exponential Moving Average

In [ ]:

data_SPY = yf.download("SPY", start = "2001-01-01", end = "2023-01-01")


print(data_SPY)

[*********************100%***********************]  1 of 1 completed
                  Open        High         Low       Close   Adj Close  \
Date                                                                     
2001-01-02  132.000000  132.156250  127.562500  128.812500   84.662971   
2001-01-03  128.312500  136.000000  127.656250  135.000000   88.729744   
2001-01-04  134.937500  135.468750  133.000000  133.546875   87.774704   
2001-01-05  133.468750  133.625000  129.187500  129.187500   84.909470   
2001-01-08  129.875000  130.187500  127.687500  130.187500   85.566696   
...                ...         ...         ...         ...         ...   
2022-12-23  379.649994  383.059998  378.029999  382.910004  380.042480   
2022-12-27  382.790009  383.149994  379.649994  381.399994  378.543793   
2022-12-28  381.329987  383.390015  376.420013  376.660004  373.839264   
2022-12-29  379.630005  384.350006  379.079987  383.440002  380.568512   
2022-12-30  380.640015  382.579987  378.429

# Implementing Pandas.

In [ ]:
df = pd.DataFrame(data_SPY)
df

Open        High         Low       Close   Adj Close  \
Date                                                                     
2001-01-02  132.000000  132.156250  127.562500  128.812500   84.662971   
2001-01-03  128.312500  136.000000  127.656250  135.000000   88.729744   
2001-01-04  134.937500  135.468750  133.000000  133.546875   87.774704   
2001-01-05  133.468750  133.625000  129.187500  129.187500   84.909470   
2001-01-08  129.875000  130.187500  127.687500  130.187500   85.566696   
...                ...         ...         ...         ...         ...   
2022-12-23  379.649994  383.059998  378.029999  382.910004  380.042480   
2022-12-27  382.790009  383.149994  379.649994  381.399994  378.543793   
2022-12-28  381.329987  383.390015  376.420013  376.660004  373.839264   
2022-12-29  379.630005  384.350006  379.079987  383.440002  380.568512   
2022-12-30  380.640015  382.579987  378.429993  382.429993  379.566040   

              Volume  
Date                  
2001-01-02   8737500  
2001-01-03  19431600  
2001-01-04   9219000  
2001-01-05  12911400  
2001-01-08   6625300  
...              ...  
2022-12-23  59857300  
2022-12-27  51638200  
2022-12-28  70911500  
2022-12-29  66970900  
2022-12-30  84022200  

[5535 rows x 6 columns]

In [ ]:
df.shape

(5535, 6)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 5535 entries, 2001-01-02 to 2022-12-30
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Open       5535 non-null   float64
 1   High       5535 non-null   float64
 2   Low        5535 non-null   float64
 3   Close      5535 non-null   float64
 4   Adj Close  5535 non-null   float64
 5   Volume     5535 non-null   int64  
dtypes: float64(5), int64(1)
memory usage: 302.7 KB


In [ ]:
#if columns are cut off..

pd.set_option('display.max_columns', 6)
df

Open        High         Low       Close   Adj Close  \
Date                                                                     
2001-01-02  132.000000  132.156250  127.562500  128.812500   84.662971   
2001-01-03  128.312500  136.000000  127.656250  135.000000   88.729744   
2001-01-04  134.937500  135.468750  133.000000  133.546875   87.774704   
2001-01-05  133.468750  133.625000  129.187500  129.187500   84.909470   
2001-01-08  129.875000  130.187500  127.687500  130.187500   85.566696   
...                ...         ...         ...         ...         ...   
2022-12-23  379.649994  383.059998  378.029999  382.910004  380.042480   
2022-12-27  382.790009  383.149994  379.649994  381.399994  378.543793   
2022-12-28  381.329987  383.390015  376.420013  376.660004  373.839264   
2022-12-29  379.630005  384.350006  379.079987  383.440002  380.568512   
2022-12-30  380.640015  382.579987  378.429993  382.429993  379.566040   

              Volume  
Date                  
2001-01-02   8737500  
2001-01-03  19431600  
2001-01-04   9219000  
2001-01-05  12911400  
2001-01-08   6625300  
...              ...  
2022-12-23  59857300  
2022-12-27  51638200  
2022-12-28  70911500  
2022-12-29  66970900  
2022-12-30  84022200  

[5535 rows x 6 columns]

In [ ]:
df.head()   #this prints first 5 rows

Open       High        Low       Close  Adj Close    Volume
Date                                                                        
2001-01-02  132.00000  132.15625  127.56250  128.812500  84.662971   8737500
2001-01-03  128.31250  136.00000  127.65625  135.000000  88.729744  19431600
2001-01-04  134.93750  135.46875  133.00000  133.546875  87.774704   9219000
2001-01-05  133.46875  133.62500  129.18750  129.187500  84.909470  12911400
2001-01-08  129.87500  130.18750  127.68750  130.187500  85.566696   6625300

In [ ]:
df.head(10) #prints first 10 rows

Open        High         Low       Close  Adj Close  \
Date                                                                    
2001-01-02  132.000000  132.156250  127.562500  128.812500  84.662971   
2001-01-03  128.312500  136.000000  127.656250  135.000000  88.729744   
2001-01-04  134.937500  135.468750  133.000000  133.546875  87.774704   
2001-01-05  133.468750  133.625000  129.187500  129.187500  84.909470   
2001-01-08  129.875000  130.187500  127.687500  130.187500  85.566696   
2001-01-09  131.046875  131.500000  129.421875  129.843750  85.340752   
2001-01-10  129.000000  132.125000  128.812500  132.125000  86.840141   
2001-01-11  131.093750  133.484375  131.093750  132.250000  86.922325   
2001-01-12  132.687500  133.718750  131.281250  132.000000  86.757980   
2001-01-16  132.000000  133.187500  131.515625  132.843750  87.312538   

              Volume  
Date                  
2001-01-02   8737500  
2001-01-03  19431600  
2001-01-04   9219000  
2001-01-05  12911400  
2001-01-08   6625300  
2001-01-09   5702400  
2001-01-10   8746100  
2001-01-11   7245100  
2001-01-12   7244000  
2001-01-16   8542200

In [ ]:
df.tail()   #last 5 rows

Open        High         Low       Close   Adj Close  \
Date                                                                     
2022-12-23  379.649994  383.059998  378.029999  382.910004  380.042480   
2022-12-27  382.790009  383.149994  379.649994  381.399994  378.543793   
2022-12-28  381.329987  383.390015  376.420013  376.660004  373.839264   
2022-12-29  379.630005  384.350006  379.079987  383.440002  380.568512   
2022-12-30  380.640015  382.579987  378.429993  382.429993  379.566040   

              Volume  
Date                  
2022-12-23  59857300  
2022-12-27  51638200  
2022-12-28  70911500  
2022-12-29  66970900  
2022-12-30  84022200

In [ ]:
df['Close']   #or df.Close

Date
2001-01-02    128.812500
2001-01-03    135.000000
2001-01-04    133.546875
2001-01-05    129.187500
2001-01-08    130.187500
                 ...    
2022-12-23    382.910004
2022-12-27    381.399994
2022-12-28    376.660004
2022-12-29    383.440002
2022-12-30    382.429993
Name: Close, Length: 5535, dtype: float64

In [ ]:
type(df['Close'])

pandas.core.series.Series

In [ ]:
df[['Close', 'Open']]

Close        Open
Date                              
2001-01-02  128.812500  132.000000
2001-01-03  135.000000  128.312500
2001-01-04  133.546875  134.937500
2001-01-05  129.187500  133.468750
2001-01-08  130.187500  129.875000
...                ...         ...
2022-12-23  382.910004  379.649994
2022-12-27  381.399994  382.790009
2022-12-28  376.660004  381.329987
2022-12-29  383.440002  379.630005
2022-12-30  382.429993  380.640015

[5535 rows x 2 columns]

In [ ]:
df.columns

Index(['Open', 'High', 'Low', 'Close', 'Adj Close', 'Volume'], dtype='object')

In [ ]:
df.iloc[0]

Open         1.320000e+02
High         1.321562e+02
Low          1.275625e+02
Close        1.288125e+02
Adj Close    8.466297e+01
Volume       8.737500e+06
Name: 2001-01-02 00:00:00, dtype: float64

In [ ]:
df.iloc[[0, 1]]

Open       High        Low     Close  Adj Close    Volume
Date                                                                     
2001-01-02  132.0000  132.15625  127.56250  128.8125  84.662971   8737500
2001-01-03  128.3125  136.00000  127.65625  135.0000  88.729744  19431600

In [ ]:
df.iloc[[0, 1], 2]  #This printed Low

Date
2001-01-02    127.56250
2001-01-03    127.65625
Name: Low, dtype: float64

iloc searches by integer location,

with loc, we search by label.

In [ ]:
df.loc['2001-01-02']

Open         1.320000e+02
High         1.321562e+02
Low          1.275625e+02
Close        1.288125e+02
Adj Close    8.466297e+01
Volume       8.737500e+06
Name: 2001-01-02 00:00:00, dtype: float64

In [ ]:
df.loc[['2001-01-02', '2001-01-03']]

Open       High        Low     Close  Adj Close    Volume
Date                                                                     
2001-01-02  132.0000  132.15625  127.56250  128.8125  84.662971   8737500
2001-01-03  128.3125  136.00000  127.65625  135.0000  88.729744  19431600

In [ ]:
df.loc[['2001-01-02', '2001-01-03'], ['Close', 'Open']]   #['Close', 'Open'] order is respected

Close      Open
Date                          
2001-01-02  128.8125  132.0000
2001-01-03  135.0000  128.3125

In [ ]:
df['Close']

Date
2001-01-02    128.812500
2001-01-03    135.000000
2001-01-04    133.546875
2001-01-05    129.187500
2001-01-08    130.187500
                 ...    
2022-12-23    382.910004
2022-12-27    381.399994
2022-12-28    376.660004
2022-12-29    383.440002
2022-12-30    382.429993
Name: Close, Length: 5535, dtype: float64

In [ ]:
df['Close'].value_counts()

131.970001    6
114.199997    5
82.750000     4
126.660004    4
112.580002    4
             ..
109.430000    1
110.989998    1
109.940002    1
111.300003    1
382.429993    1
Name: Close, Length: 4762, dtype: int64

In [ ]:
df.loc['2001-01-02', 'Close']

128.8125

In [ ]:
df.loc[['2001-01-02','2001-01-03','2001-01-04','2001-01-05'], 'Close']

Date
2001-01-02    128.812500
2001-01-03    135.000000
2001-01-04    133.546875
2001-01-05    129.187500
Name: Close, dtype: float64

In [ ]:
df.loc['2001-01-02':'2001-01-12', 'Close':'Volume']    #Utilizing slicing

Close  Adj Close    Volume
Date                                       
2001-01-02  128.812500  84.662971   8737500
2001-01-03  135.000000  88.729744  19431600
2001-01-04  133.546875  87.774704   9219000
2001-01-05  129.187500  84.909470  12911400
2001-01-08  130.187500  85.566696   6625300
2001-01-09  129.843750  85.340752   5702400
2001-01-10  132.125000  86.840141   8746100
2001-01-11  132.250000  86.922325   7245100
2001-01-12  132.000000  86.757980   7244000

In [ ]:
df.set_index('Close')

Open        High         Low   Adj Close    Volume
Close                                                               
128.812500  132.000000  132.156250  127.562500   84.662971   8737500
135.000000  128.312500  136.000000  127.656250   88.729744  19431600
133.546875  134.937500  135.468750  133.000000   87.774704   9219000
129.187500  133.468750  133.625000  129.187500   84.909470  12911400
130.187500  129.875000  130.187500  127.687500   85.566696   6625300
...                ...         ...         ...         ...       ...
382.910004  379.649994  383.059998  378.029999  380.042480  59857300
381.399994  382.790009  383.149994  379.649994  378.543793  51638200
376.660004  381.329987  383.390015  376.420013  373.839264  70911500
383.440002  379.630005  384.350006  379.079987  380.568512  66970900
382.429993  380.640015  382.579987  378.429993  379.566040  84022200

[5535 rows x 5 columns]

In [ ]:
df.index

DatetimeIndex(['2001-01-02', '2001-01-03', '2001-01-04', '2001-01-05',
               '2001-01-08', '2001-01-09', '2001-01-10', '2001-01-11',
               '2001-01-12', '2001-01-16',
               ...
               '2022-12-16', '2022-12-19', '2022-12-20', '2022-12-21',
               '2022-12-22', '2022-12-23', '2022-12-27', '2022-12-28',
               '2022-12-29', '2022-12-30'],
              dtype='datetime64[ns]', name='Date', length=5535, freq=None)

In [ ]:
df.set_index('Close', inplace=True)   #Sets index into 'Close'

In [ ]:
df.index

Float64Index([          128.8125,              135.0,         133.546875,
                        129.1875,           130.1875,          129.84375,
                         132.125,             132.25,              132.0,
                       132.84375,
              ...
               383.2699890136719,  380.0199890136719,  380.5400085449219,
               386.2300109863281,  380.7200012207031,  382.9100036621094,
               381.3999938964844,  376.6600036621094, 383.44000244140625,
              382.42999267578125],
             dtype='float64', name='Close', length=5535)

In [ ]:
df.loc[132.0, 'Volume']


7244000.0

In [ ]:
df.reset_index(inplace=True)    #Resets the index, if run multiple times it assins new numerical index to it
df                              # a.k.a. duplicates. You can use 'if' to check beforehand.

ValueError: ignored

In [ ]:
if df.index.name is not None:
  df.reset_index(inplace=True)

In [ ]:
df['Close'] >= 150    #Can filter out or detect the point of change.

Date
2001-01-02    False
2001-01-03    False
2001-01-04    False
2001-01-05    False
2001-01-08    False
              ...  
2022-12-23     True
2022-12-27     True
2022-12-28     True
2022-12-29     True
2022-12-30     True
Name: Close, Length: 5535, dtype: bool

In [ ]:
filt = (df['Close'] >= 150)
df[filt]

Open        High         Low       Close   Adj Close  \
Date                                                                     
2007-05-03  149.970001  150.399994  149.729996  150.350006  109.577545   
2007-05-04  150.750000  151.119995  150.220001  150.919998  109.992973   
2007-05-07  150.880005  151.199997  150.809998  150.949997  110.014870   
2007-05-08  150.580002  150.919998  150.130005  150.750000  109.869026   
2007-05-09  150.639999  152.820007  150.369995  151.160004  110.167938   
...                ...         ...         ...         ...         ...   
2022-12-23  379.649994  383.059998  378.029999  382.910004  380.042480   
2022-12-27  382.790009  383.149994  379.649994  381.399994  378.543793   
2022-12-28  381.329987  383.390015  376.420013  376.660004  373.839264   
2022-12-29  379.630005  384.350006  379.079987  383.440002  380.568512   
2022-12-30  380.640015  382.579987  378.429993  382.429993  379.566040   

               Volume  
Date                   
2007-05-03   86569700  
2007-05-04   96409000  
2007-05-07   63461400  
2007-05-08   80584000  
2007-05-09  102070100  
...               ...  
2022-12-23   59857300  
2022-12-27   51638200  
2022-12-28   70911500  
2022-12-29   66970900  
2022-12-30   84022200  

[2590 rows x 6 columns]

In [ ]:
df.loc[filt, 'Close']

Date
2007-05-03    150.350006
2007-05-04    150.919998
2007-05-07    150.949997
2007-05-08    150.750000
2007-05-09    151.160004
                 ...    
2022-12-23    382.910004
2022-12-27    381.399994
2022-12-28    376.660004
2022-12-29    383.440002
2022-12-30    382.429993
Name: Close, Length: 2590, dtype: float64

In [ ]:
filt = (df['Close'] >= 150) & (df['Close'] <= 250)
df.loc[filt, 'Close']

Date
2007-05-03    150.350006
2007-05-04    150.919998
2007-05-07    150.949997
2007-05-08    150.750000
2007-05-09    151.160004
                 ...    
2020-03-23    222.949997
2020-03-24    243.149994
2020-03-25    246.789993
2020-04-01    246.149994
2020-04-03    248.190002
Name: Close, Length: 1279, dtype: float64

In [ ]:
filt = (df['Close'] >= 150) | (df['Close'] <= 250)
df.loc[filt, 'Close']

Date
2001-01-02    128.812500
2001-01-03    135.000000
2001-01-04    133.546875
2001-01-05    129.187500
2001-01-08    130.187500
                 ...    
2022-12-23    382.910004
2022-12-27    381.399994
2022-12-28    376.660004
2022-12-29    383.440002
2022-12-30    382.429993
Name: Close, Length: 5535, dtype: float64

In [ ]:
filt = (df['Close'] >= 150) & (df['Close'] <= 250)
df.loc[~filt, 'Close']

Date
2001-01-02    128.812500
2001-01-03    135.000000
2001-01-04    133.546875
2001-01-05    129.187500
2001-01-08    130.187500
                 ...    
2022-12-23    382.910004
2022-12-27    381.399994
2022-12-28    376.660004
2022-12-29    383.440002
2022-12-30    382.429993
Name: Close, Length: 4256, dtype: float64

since ~ was added before 'filt', the difference in the result could be seen through the length. One with the ~ before 'filt' has 4256 length.